File: 334. Increasing Triplet Subsequence.cpp
334. Increasing Triplet Subsequenceclass Solution {
public:
    // dp[i]: length of Longest Increasing Subarray ending at i
    // O(N2)time, O(N)space, TLE for last big data case;
    bool increasingTriplet(vector<int>& nums) {
        if(nums.empty()) return false;
        vector<int> len(nums.size(), 1);  // base case
        for(int i = 1; i < nums.size(); i++){
            for(int j = 0; j < i; j++){
                if(nums[i] > nums[j])
                    len[i] = max(len[i], len[j] + 1);
            }
            if(len[i] == 3) return true;
        }
        return false;
    }

    // binary search: O(NlogN)time, O(N)space, 33.25%, ok
    // keep an increasing list, replace the smallest elem that is larger than nums[i]
    bool increasingTriplet(vector<int>& nums) {
        if(nums.empty()) return false;
        vector<int> list(nums.size(), INT_MAX);  // initialize to max
        for(int i = 0; i < nums.size(); i++){
            // binary search
            int head = 0, tail = i;
            while(head < tail){
                int mid = head + (tail - head) / 2;
                if(list[mid] > nums[i]) tail = mid;
                else if(list[mid] < nums[i]) head = mid + 1;
                else{ head = mid; break;}
            }
            list[head] = nums[i];
            if(head + 1 == 3) return true;
        }
        return false;
    }

};

File: 334. Increasing Triplet Subsequence.py
class Solution(object):
    
    # brute force: O(N3) time, O(1) space, TLE
    def increasingTriplet(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        if len(nums) < 3:
            return False
        for i in xrange(len(nums)):
            for j in xrange(i + 1, len(nums)):
                for k in xrange(j + 1, len(nums)):
                    if nums[i] < nums[j] < nums[k]:
                        return True
        return False
    
    # stack: O(N2) time, O(1) space, TLE
    def increasingTriplet(self, nums):
        for i in xrange(len(nums) - 2):
            stack = [nums[i]]
            for j in xrange(i + 1, len(nums)):
                if len(stack) > 1 and stack[-1] > nums[j] and stack[-2] < nums[j]:
                    stack.pop()
                    stack.append(nums[j])
                elif nums[j] > stack[-1]:
                    stack.append(nums[j])
                if len(stack) >= 3:
                    return True
        return False
    
    # !! dynamic programming: dp[i] = longest length of increasing subsequece ending at i
    # O(N2) time, O(N) space, TLE
    def increasingTriplet(self, nums):
        if len(nums) < 3:
            return False
        dp = [1 for i in range(len(nums))]
        for end in xrange(1, len(nums)):
            for start in xrange(end):
                if nums[end] > nums[start]:
                    dp[end] = max(dp[end], dp[start] + 1)
            if dp[end] >= 3:
                return True
        return False
    
    # !!! two marker: O(N) time, O(1) space
    def increasingTriplet(self, nums):
        if len(nums) < 3:
            return False
        min1, min2 = float('inf'), float('inf')
        for num in nums:
            if num <= min1:
                min1 = num
            elif num <= min2:
                min2 = num
            else:
                return True
        return False
        